document/appendix/implementation.rst

```diff
@@ -0,0 +1,130 @@
+* the size of any :ref:`section <binary-section>`
+* the size of an individual :ref:`function's code <binary-func>`
+* the number of :ref:`sections <binary-section>`
+
+
+.. _impl-text:
+.. index:: text format, source text, token, identifier, character, unicode
+
+Text Format
+...........
+
+For a module given in :ref:`text format <text>`, additional limitations may be imposed on the following dimensions:
+
+* the size of the :ref:`source text <source>`
+* the size of any syntactic element
+* the size of an individual :ref:`token <text-token>`
+* the nesting depth of :ref:`folded instructions <text-foldedinstr>`
+* the length of symbolic :ref:`identifiers <text-id>`
+* the range of literal :ref:`characters <text-char>` (code points) allowed in the :ref:`source text <source>`
+
+
+.. _impl-valid:
+.. index:: validation, function
+
+Validation
+~~~~~~~~~~
+
+An implementation may defer :ref:`validation <valid>` of individual :ref:`functions <syntax-func>` until they are first :ref:`invoked <exec-invoke>`.
+
+If a function turns out to be invalid, then the invocation, and every consecutive call to the same function, results in a :ref:`trap <trap>`.
+
+.. note::
+   This is to allow implementations to use interpretation or just-in-time compilation for functions.
+   The function must still be fully validated before execution of its body begins.
+
+
+.. _impl-exec:
+.. index:: execution, module instance, function instance, table instance, memory instance, global instance, allocation, frame, label, value
+
+Execution
+~~~~~~~~~
+
+Restrictions on the following dimensions may be imposed during :ref:`execution <exec>` of a WebAssembly program:
+
+* the number of allocated :ref:`module instances <syntax-moduleinst>`
+* the number of allocated :ref:`function instances <syntax-funcinst>`
+* the number of allocated :ref:`table instances <syntax-tableinst>`
+* the number of allocated :ref:`memory instances <syntax-meminst>`
+* the number of allocated :ref:`global instances <syntax-globalinst>`
+* the size of a :ref:`table instance <syntax-tableinst>`
+* the size of a :ref:`memory instance <syntax-meminst>`
+* the number of :ref:`frames <syntax-frame>` on the :ref:`stack <syntax-stack>`
+* the number of :ref:`labels <syntax-label>` on the :ref:`stack <syntax-stack>`
+* the number of :ref:`values <syntax-val>` on the :ref:`stack <syntax-stack>`
+
+If the runtime limits of an implementation are exceeded during execution of a computation,
+then it may terminate that computation by causing a trap or reporting an embedder-specific error to the invoking code.
+
+Some of the above limits may already be verified during instantiation, in which case an implementation may report exceedance in the same manner as for :ref:`syntactic limits <impl-syntax>`.
+
+.. note::
+   Concrete limits are usually not fixed but may be dependent on specifics, interdependent, vary over time, or depend on other implementation- or embedder-specific variables.
```

Comment:
Generalised to "implementation- or embedder-specific situations or events". Is that broad enough? I intentionally avoided being more specific or even enumerating examples, since it seems futile trying to compile an open-ended list.